src/main.rs

```diff
@@ -118,6 +124,128 @@ fn main() {
+        web_sys::console::log_1(&js_sys::JsString::from(report.to_string()));
+    }));
+
+    // Redirect tracing to console.log and friends:
+    tracing_wasm::set_as_global_default();
+
+    // Redirect log (currently used by egui) to tracing
+    tracing_log::LogTracer::init().expect("failed to initialize tracing-log");
+
+    let window = web_sys::window().expect("could not get `window` object (make sure you're running this in the main thread of a web browser)");
+    let prefers_color_scheme_dark = window
+        .match_media("(prefers-color-scheme: dark)")
+        .unwrap()
+        .map(|x| x.matches());
+
+    let canvas = window
+        .document()
+        .expect("could not get `window.document` object (make sure you're running this in a web browser)")
+        .get_element_by_id(CANVAS_ID)
+        .expect(format!("could not find HTML element with ID '{CANVAS_ID}'").as_str())
+        .unchecked_into::<web_sys::HtmlCanvasElement>();
+    let offscreen_canvas = canvas
+        .transfer_control_to_offscreen()
+        .expect("could not transfer canvas control to offscreen");
+
+    let (filesystem_tx, filesystem_rx) = mpsc::unbounded_channel();
+    let (output_tx, output_rx) = mpsc::unbounded_channel();
+    let (event_tx, event_rx) = mpsc::unbounded_channel();
+    let (custom_event_tx, custom_event_rx) = mpsc::unbounded_channel();
+
+    filesystem::web::setup_main_thread_hooks(filesystem_rx);
+    luminol::web::web_worker_runner::setup_main_thread_hooks(
+        canvas,
+        event_tx.clone(),
+        custom_event_tx.clone(),
+        output_rx,
+    );
+
+    *WORKER_DATA.borrow_mut() = Some(WorkerData {
+        audio: luminol::audio::Audio::default().into(),
+        prefers_color_scheme_dark,
+        filesystem_tx,
+        output_tx,
+        event_rx,
+        custom_event_rx,
+    });
+
+    let mut worker_options = web_sys::WorkerOptions::new();
+    worker_options.name("luminol-primary");
+    worker_options.type_(web_sys::WorkerType::Module);
+    let worker = web_sys::Worker::new_with_options("/worker.js", &worker_options)
+        .expect("failed to spawn web worker");
+
+    let message = js_sys::Array::new();
+    message.push(&JsValue::from("init"));
+    message.push(&wasm_bindgen::memory());
+    message.push(&offscreen_canvas);
+    let transfer = js_sys::Array::new();
+    transfer.push(&offscreen_canvas);
+    worker
+        .post_message_with_transfer(&message, &transfer)
+        .expect("failed to post message to web worker");
+}
+
+#[cfg(target_arch = "wasm32")]
+#[wasm_bindgen]
+pub async fn luminol_worker_start(canvas: web_sys::OffscreenCanvas) {
+    let WorkerData {
+        audio,
+        prefers_color_scheme_dark,
+        filesystem_tx,
+        output_tx,
+        event_rx,
+        custom_event_rx,
+    } = WORKER_DATA.borrow_mut().take().unwrap();
+
+    filesystem::web::FileSystem::setup_filesystem_sender(filesystem_tx);
+
+    let web_options = eframe::WebOptions::default();
```

Comment:
I had to make a custom app runner for web builds and I don't think I included WebGL support. I don't think Luminol would even run in WebGL right now because it has a texture size limit of 2048 instead of 8192. But I could still put WebGL support into the app runner if you want.

Reply:
Ah, alright! You don't need to. It means I'll have to test this on windows (chrome's linux webgpu support is half-baked) but that's ok!